File: pentevo/tools/fsplit/src/fsplit/fsplit.cpp
// fsplit.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"

#define MAX_SPLIT 100

int _tmain(int argc, _TCHAR* argv[])
{
	if ((argc < 3) || (argc > (MAX_SPLIT + 2)))
	{
		printf("File splitter by TS-Labs\n");
		printf("Usage: fsplit.exe <input> <piece 1 size> .. <piece N size>\n");
		return 1;
	}

	FILE* f_in = _wfopen (argv[1], L"rb");
	if (!f_in)
	{
		printf ("Input file error!");
		return 1;
	}

    struct _stat st;
    _wstat(argv[1], &st);
    int sz = st.st_size;
    void *buf = malloc(sz);
    if (!buf)
    {
        printf ("Memory alloc error!");
        return 3;
    }

    int ssz[MAX_SPLIT];
    int n = (argc - 2);
    for (int i = 0; i < n; i++)
    {
        ssz[i] = min(sz, _wtoi(argv[i + 2]));
        sz -= ssz[i];
        if (!sz)
            break;
    }
        
    if (sz)
        ssz[n++] = sz;

    for (int i = 0; i < n; i++)
	{
        _TCHAR fout[256];
        _TCHAR t[256];
        wcscpy(fout, argv[1]);
        wcscat(fout, L".");
        wcscat(fout, _itow(i, t, 10));

        FILE *f_out = _wfopen(fout, L"wb");
        if (!f_out)
        {
            printf ("Output file error!");
            return 2;
        }

        if (fread(buf, 1, ssz[i], f_in) != ssz[i])
        {
            printf ("Read file error!");
            return 4;
        }

        if (fwrite(buf, 1, ssz[i], f_out) != ssz[i])
        {
            printf ("Write file error!");
            return 5;
        }

        fclose(f_out);
    }

    fclose(f_in);

    printf("DONE!\n");

	return 0;
}

File: pentevo/tools/fsplit/fsplit.py
import sys
import os
from pathlib import Path

MAX_SPLIT = 100

def main():
    argc = len(sys.argv)
    if argc < 3 or argc > (MAX_SPLIT + 2):
        print("File splitter by TS-Labs")
        print("Usage: fsplit.py <input> <piece 1 size> .. <piece N size>")
        return 1

    input_file = sys.argv[1]
    try:
        with open(input_file, "rb") as f_in:
            # Get file size
            st = os.stat(input_file)
            sz = st.st_size
            buf = f_in.read(sz)  # Read entire file into memory
            if len(buf) != sz:
                print("Read file error!")
                return 4
    except IOError:
        print("Input file error!")
        return 1

    # Parse piece sizes
    ssz = [0] * MAX_SPLIT
    n = argc - 2
    for i in range(n):
        try:
            ssz[i] = min(sz, int(sys.argv[i + 2]))
            sz -= ssz[i]
            if sz == 0:
                break
        except ValueError:
            print("Invalid piece size!")
            return 1

    if sz:
        ssz[n] = sz
        n += 1

    # Split file into pieces
    offset = 0
    for i in range(n):
        output_file = f"{input_file}.{i}"
        try:
            with open(output_file, "wb") as f_out:
                bytes_to_write = ssz[i]
                if f_out.write(buf[offset:offset + bytes_to_write]) != bytes_to_write:
                    print("Write file error!")
                    return 5
                offset += bytes_to_write
        except IOError:
            print("Output file error!")
            return 2

    print("DONE!")
    return 0

if __name__ == "__main__":
    sys.exit(main())
